src/scanner.cc

```diff
@@ -486,16 +486,15 @@ static void consume_until_v2(string target, State &state) {
   }
 }
 
-static function<u32string(State &)> read_string(Peek pred) {
-  return [=](State & state) {
-    u32string s;
-    consume_while([&](uint32_t c) {
-        auto res = pred(c);
-        if (res) s += static_cast<uint32_t>(c);
-        return res;
-    })(state);
-    return s;
-  };
+static u32string read_string(bool (*cond)(uint32_t), State &state) {
```

Comment:
Use a templated function instead of a function pointer. This might help the optimizer inline it.

```
template <typename F>
static u32string read_string(F cond, State &state) {
```
(and use `&` or `&&` after F to avoid a copy?)